### PR TITLE
Improved cnn encoder decoder

### PR DESCRIPTION
**Pull request by jemrobinson**
Tested various potential improvements to the CNNEncoder and CNNDecoder. Many of these did not work at all, but the best performing (after 100 epochs of training with 1 year) is the following:

**Encoder**
- Resize with interpolation (if needed) so that the shape after convolution will *exactly* match the desired output shape
- n_layers of size-reducing convolutional blocks

**Decoder**
- Increase size with interpolation (if needed) so that the shape after convolution will be as-close-as-possible to the output shape
- n_layers of size-increasing convolutional blocks
- Decrease size with interpolation (if needed) so that the shape *exactly* matches the output shape
- Convolve to number of output channels (if needed) so that the number of channels *exactly* matches the output channels

## Results

Running with default settings uses the following layers:

```
CNNEncoder (osisaf-south) with 3 layers
- ResizingInterpolation from (432, 432) to (1024, 1024)
- ConvBlockDownsample (ReLU, 3) with 1 channels
- ConvBlockDownsample (ReLU, 3) with 2 channels
- ConvBlockDownsample (ReLU, 3) with 4 channels
CNNEncoder (era5) with 3 layers
- ResizingInterpolation from (181, 720) to (1024, 1024)
- ConvBlockDownsample (ReLU, 3) with 27 channels
- ConvBlockDownsample (ReLU, 3) with 54 channels
- ConvBlockDownsample (ReLU, 3) with 108 channels
CNNDecoder (osisaf-south) with 3 layers
- ConvBlockUpsample (ReLU, 3) with 224 channels
- ConvBlockUpsample (ReLU, 3) with 112 channels
- ConvBlockUpsample (ReLU, 3) with 56 channels
- ResizingInterpolation from (1024, 1024) to (432, 432)
- Channel convolution from 28 to 1
```

**CNN-UNet-CNN**
cnn-11-100epochs-f5c4547

**CNN-ViT-CNN**
media_images_sea-ice_concentration-static-maps_0_914c2320bbb8666d51c4

Note that there's still a bit of aliasing. I think @louisavz might have some thoughts about how to improve this?

## Results with anti-aliasing

**CNN-UNet-CNN**
media_images_sea-ice_concentration-static-maps_0_4c56074e6e213103f921

**CNN-ViT-CNN**
media_images_sea-ice_concentration-static-maps_0_a0cd9455d817619c384b

### CODE REVIEW

**Comment by louisavz** [comment]
Hi @jemrobinson, regarding aliasing and the blockiness of the output [`nn.functional.interpolate`](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html) has an optional `antialias=False` might worth looking into as an optional input argument (currently not an option in [L25](https://github.com/alan-turing-institute/ice-station-zebra/blob/1542b6c030b030d0bfa5e1e266009a7366c5e251/ice_station_zebra/models/common/resizing_interpolation.py#L25C16-L27C10) of `resizing_interpolation.py`). Another note is I used resize (with Gaussian) for every up convolutional layers. In Tensorflow:

```
tf.image.resize(
            inputs,
            size=[out_height, out_width],
            method=tf.image.ResizeMethod.GAUSSIAN
)
```

File: ice_station_zebra/models/decoders/cnn_decoder.py [hunk 1]
```diff
@@ -36,31 +41,72 @@ def __init__(
...
+            max(minimal_input_shape[0], self.data_space_in.shape[0]),
+            max(minimal_input_shape[1], self.data_space_in.shape[1]),
+        )
+        if shape != self.data_space_in.shape:
+            layers.append(ResizingInterpolation(shape))
+            logger.debug(
+                "- ResizingInterpolation from %s to %s",
+                self.data_space_in.shape,
+                minimal_input_shape,
+            )
```
Comment (louisavz) [inline ice_station_zebra/models/decoders/cnn_decoder.py:75]: I'm confused with why this is necessary. There is already a ResizingInterpolation in place at the end, why are we resizing up to minimal_input_shape before starting the conv layers? Why is a minimal_input_shape necessary?

**Comment by jemrobinson** [comment]
I think anti-aliasing has helped a bit here.

File: ice_station_zebra/models/encoders/cnn_encoder.py [hunk 2]
```diff
@@ -33,28 +37,38 @@ def __init__(
...
-
-        # Set the spatial dimensions as required by the number of convolutional layers
+        # If necessary, apply a convolutional resizing to get the correct input dimensions
         initial_required_shape = (
             self.data_space_out.shape[0] * (2**n_layers),
             self.data_space_out.shape[1] * (2**n_layers),
         )
-        layers.append(ResizingInterpolation(initial_required_shape))
+        if self.data_space_in.shape != initial_required_shape:
+            layers.append(ResizingInterpolation(initial_required_shape))
```
Comment (louisavz) [inline ice_station_zebra/models/encoders/cnn_encoder.py:48]: I think this is similar to the decoder's minimal_input_shape question above. Why force the initial_required_shape to be a multiple of 2**n_layers. Are you avoiding the use of padding (or other options) for the convolution?

**Review by louisavz — COMMENTED** [review]
Hi @jemrobinson, I have some questions and rather than me guessing the intents, perhaps you could help me understand better.

**Comment by louisavz** [comment]
> I think anti-aliasing has helped a bit here.

Yea, looks like it, at least, for the background. Not sure what happened with the ViT's. Well done!

**Comment by louisavz** [comment]
Although the range is a bit concerning...

File: ice_station_zebra/models/decoders/cnn_decoder.py (hunk 1 above)
Comment (jemrobinson) [reply ice_station_zebra/models/decoders/cnn_decoder.py:75]: The plan was to do the crude size-increasing interpolation before the convolutional layers, so that the "added" information will get smoothed out by the convolution. This means that the final resizing interpolation layer should always be *size-decreasing* which is (hopefully) less likely to produce artifacts. I'm not sure whether this is actually true though - I can try with just the final resizing layer.

File: ice_station_zebra/models/encoders/cnn_encoder.py (hunk 2 above)
Comment (jemrobinson) [reply ice_station_zebra/models/encoders/cnn_encoder.py:48]: This is the shape that will give the correct output shape after convolution, since the ConvBlockDownsample layers each reduce (H, W) by 2 while doubling C. I'll see whether doing a simple resize at the end of the process performs well.

File: ice_station_zebra/models/common/resizing_interpolation.py [hunk 3]
```diff
@@ -23,5 +23,9 @@ def __init__(self, output_shape: Sequence[int]) -> None:
 
     def forward(self, x: TensorNCHW) -> TensorNCHW:
         return nn.functional.interpolate(
-            x, size=self.output_shape, mode="bilinear", align_corners=True
+            x,
+            size=self.output_shape,
+            mode="bilinear",
+            align_corners=True,
+            antialias=True,
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply ice_station_zebra/models/common/resizing_interpolation.py]: Done in 88017c0

File: ice_station_zebra/models/decoders/cnn_decoder.py (hunk 1 above)
Comment (jemrobinson) [reply ice_station_zebra/models/decoders/cnn_decoder.py:77]: Would it make sense to do e.g. a centre-crop instead of a resize for the final size-reducing step?

Comment (louisavz) [reply ice_station_zebra/models/decoders/cnn_decoder.py:77]: If we do a centre-crop, wouldn't that make the region out of scale? Each pixel has distance attached to it. If our input is say 432x432, then the output should be 432x432. If the last layer output is say 562x562, the information is contained inside those 562x562 pixels representing information for 432x432, and needs to be shaped back accordingly. I'm not sure if cropping makes sense. Shouldn't we be making the final conv layer outputs the expected shape so resize is not necessary?

**Comment by louisavz** [comment]
Regarding range, I need to dive in the code, right now I'm only seeing snippets and trying to take a best guess. Let's keep to what we discussed on slack about sigmoid, as I'm worry that adding sigmoid might mask potential issues and we might forget to dig in deeper. For the ViT, the range looks like they could be floating points range. Do you want to merge this PR as is, and try or add other methods with a new (draft) PR?

**Review by louisavz — APPROVED** [review]
LGTM and it has achieved "better than we had before".
